fiona/ogrext.pyx

```diff
@@ -1007,7 +1008,10 @@ cdef class WritingSession(Session):
                 if field_subtype != OFSTNone:
                     # subtypes are new in GDAL 2.x, ignored in 1.x
                     set_field_subtype(cogr_fielddefn, field_subtype)
-                OGR_L_CreateField(self.cogr_layer, cogr_fielddefn, 1)
+                ret = OGR_L_CreateField(self.cogr_layer, cogr_fielddefn, 1)
+                if ret != OGRERR_NONE:
+                    GDALClose(cogr_ds)
+                    raise ValueError("Failed to create field \"{}\", possibly due to mismatched encoding".format(key))
```

Comment:
@snorfalorpagus I think field creation might be too late to catch this error in some cases. We should check before line 1006 (according to https://travis-ci.org/Toblerity/Fiona/jobs/389857334#L1137) when we are creating the field definition.